Sources/SwiftDocC/Utility/ValidatedURL.swift

```diff
@@ -178,8 +178,13 @@ private extension StringProtocol {
     /// Returns a percent encoded version of the string or the original string if it is already percent encoded.
     func addingPercentEncodingIfNeeded(withAllowedCharacters allowedCharacters: CharacterSet) -> String? {
         var needsPercentEncoding: Bool {
-            for (index, character) in unicodeScalars.indexed() where !allowedCharacters.contains(character) {
+            let unicodeScalars = unicodeScalars.filter { !allowedCharacters.contains($0) }
+            for (index, character) in unicodeScalars.indexed() {
                 if character == "%" {
+                    // There's not two characters after the "%". This "%" can't represent a percent encoded character.
+                    guard index != unicodeScalars.endIndex || index != unicodeScalars.index(before: index) else {
```

Comment:
`String.UnicodeScalarView.Index` is Comparable, so this can be simplified:

```suggestion
                    guard index < self.unicodeScalars.index(self.unicodeScalars.endIndex, offsetBy: -2, limitedBy: self.unicodeScalars.startIndex) else {
```

(The `self.` prefixes can be dropped if we go back to a `for/where` loop without a previous filter; i'm just including them here for completeness.)

Reply:
I think this can also be expressed as  
```suggestion
                    guard unicodeScalars.distance(from: index, to: unicodeScalars.endIndex) <= 2 else {
```

Reply:
or ` guard index < unicodeScalars.index(after: unicodeScalars.endIndex)`

Reply:
or `guard index < unicodeScalars.index(before: unicodeScalars.endIndex)`